### PR TITLE
Add a RawOverflowPage type and start to use it

### PR DESCRIPTION
**Pull request by dcoutts**
This is preparation for integrating the `MPageAcc` type in run construction.

We add a `RawOverflowPage` type and tests, and we change `PageAcc1.singletonPage` to return `(RawPage, [RawOverflowPage])` rather than `(RawPage, RawBytes)`.

### CODE REVIEW

File: test/Test/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    label (if BA.isByteArrayPinned ba then "pinned" else "unpinned") $
+    label (if off == 0 then "offset 0" else "offset non-0") $
+
+        rawOverflowPageRawBytes (makeRawOverflowPage ba off len)
+    === RawBytes.take 4096 bytes <> padding
+  where
+    padding = RawBytes.fromVector (PV.replicate paddinglen 0)
+    paddinglen
+      | RawBytes.size bytes > 4096 = 0
+      | otherwise = 4096 - (RawBytes.size bytes `mod` 4096)
```
Comment (mheinzel) [inline test/Test/Database/LSMTree/Internal/RawOverflowPage.hs:44]: Isn't this just `4096 - RawBytes.size bytes`? You could add a `max 0`, but `replicate` should handle the negative value fine.
Actually, for a size of 4096, I think your code gives 4096 bytes of padding?

**Comment by mheinzel** [comment]
Generally looks good to me, just two details that seemed suspicious.

File: src/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPageCopy ba off len =
+    (\page -> assert (invariant page) page) $
+    RawOverflowPage 0 $ runByteArray $ do
+      mba <- newPinnedByteArray 4096
+      let suffixlen = clamp 0 4096 (min len (sizeofByteArray ba - off))
+      copyByteArray mba 0 ba off suffixlen
+      when (suffixlen < 4096) $ fillByteArray mba suffixlen 4096 0
```
[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: The second argument to `fillByteArray` is *number of bytes to fill*, not the end offset.

[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: I think we should enable  https://downloads.haskell.org/ghc/latest/docs/users_guide/debugging.html?highlight=bounds#ghc-flag--fcheck-prim-bounds on our packages, `vector` and `primitive`.

[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: > The case where len is larger than the remaining byte array is a bit suspicious to me. 

It's not. If `ByteArray` is unpinned, we must copy, as our `RawPage*`s have an invariant of being always pinned.

File: src/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+                      fillByteArray mba len (lenPages-len) 0
+                      return mba
+     in assert (sizeofByteArray ba' == lenPages) $
+        assert (lenPages `mod` 4096 == 0) $
+        rawBytesToOverflowPagesPinned 0 lenPages ba'
+
+roundUpToPageSize :: Int -> Int
+roundUpToPageSize n =
+    -- assumes pageSize = 4096:
+    ((n + 0x0fff) .&. complement 0x0fff)
```
Comment (phadej) [inline src/Database/LSMTree/Internal/RawOverflowPage.hs:157]: add a helper to `BitMath` module, and use it.

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 4]
```diff
@@ -202,7 +203,7 @@ prop_hasblobspans :: PageLogical -> Property
 prop_hasblobspans p@(PageLogical xs) =
     [ rawPageHasBlobSpanAt rawpage i /= 0 | i <- [0 .. length xs - 1] ] === blobSpans
   where
-    rawpage = fst $ toRawPage p
+    (rawpage, _overflowPages) = toRawPage p
```
Comment (phadej) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:206]: this and similar lines are unnecessary diff noise

File: src/Database/LSMTree/Internal/RawOverflowPage.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: > The second argument to fillByteArray is number of bytes to fill, not the end offset.

I am aware. And it's clear that we want to copy at most `len` bytes and we also *can't* copy more bytes than `sizeofByteArray ba - off`. The usage of `min` still seems suspicious to me.
If someone passes in a `len` that is larger than the byte array allows, that seems like a mistake from the caller to me (unless there is a legitimate use case, which I asked for). And if it's a mistake, I'd rather fail instead of silently truncating the copy and pretending to have built a page with `len` user-specified bytes.

[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: > If someone passes in a len that is larger than the byte array allows

This is an internal function.

We could add an `assert`.

[the remark this answers is not in the document]
Comment (phadej) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: ... but then we should also assert that `off` is `>=0` and `<= sizeofByteArray ba` and ... i.e. `assert (isValidSlice ba off len)`.

[the remark this answers is not in the document]
Comment (mheinzel) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: Sounds reasonable to me.

[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:94]: I also just noticed that I'd got the fill args wrong.

I agree that enabling `-fcheck-prim-bounds` in CI is a good idea.

[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs]: Note that the `primitive` package's `setByteArray` and `fillByteArray` do not currently use the ghc primop `setByteArray#` and so do not benefit from `-fcheck-prim-bounds`.

**Review by phadej — CHANGES_REQUESTED** [review]

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 5]
```diff
@@ -254,42 +255,65 @@ prop_entries_all page@(PageLogical xs) bs =
...
     page = PageLogical [(k, op, blobref)]
-    (rawpage, sfx) = toRawPage page
+    (rawpage, overflowPages) = toRawPage page
+    overflowPagesBS = RB.toByteString
+                        (mconcat (map rawOverflowPageRawBytes overflowPages))
     k' = SerialisedKey $ RB.fromByteString (unKey k)
 
-    trim :: BS.ByteString -> BS.ByteString
-    trim = BS.dropEnd sfxSize
+    compareOpEntry (Insert (Value v))
```
Comment (phadej) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:296]: add a type signature to compareOpEntry. What it does? Looks complicated, I don't udnerstand

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 6]
```diff
@@ -254,42 +255,65 @@ prop_entries_all page@(PageLogical xs) bs =
...
+      LookupEntryOverflow (Entry.Insert (SerialisedValue v')) suffixlen
+        | Nothing <- blobref
+        -> v === RB.toByteString v' <>
+                 BS.take (fromIntegral suffixlen) overflowPagesBS
+
+      LookupEntryOverflow (Entry.InsertWithBlob (SerialisedValue v')
+                                                (BlobSpan x' y'))
+                          suffixlen
+        | Just (BlobRef x y) <- blobref
+        -> v === RB.toByteString v' <>
```
Comment (phadej) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:268]: is this something a generalised `compareOpEntry` would do?

**Review by mheinzel — APPROVED** [review]
I'm happy with it now.

File: src/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+rawBytesToOverflowPagesUnpinned :: Int -> Int -> ByteArray -> [RawOverflowPage]
+rawBytesToOverflowPagesUnpinned !off !len !ba =
+    let !lenPages = roundUpToPageSize len
+        ba'       = runByteArray $ do
+                      mba <- newPinnedByteArray lenPages
+                      copyByteArray mba 0 ba off len
+                      fillByteArray mba len (lenPages-len) 0
+                      return mba
+     in assert (sizeofByteArray ba' == lenPages) $
+        assert (lenPages `mod` 4096 == 0) $
```
Comment (jorisdral) [inline src/Database/LSMTree/Internal/RawOverflowPage.hs:155]: This might fit better as an assertion inside `roundUpToPageSize`

File: test/Test/Util/RawPage.hs [hunk 8]
```diff
@@ -22,12 +24,26 @@ import           Test.Tasty.HUnit (Assertion, assertFailure)
...
+toRawPage p = (page, overflowPages)
   where
     bs = serialisePage $ encodePage p
     (pfx, sfx) = BS.splitAt 4096 bs -- hardcoded page size.
-    page = case SBS.toShort pfx of SBS.SBS ba -> makeRawPage (ByteArray ba) 0
+    page          = makeRawPageBS pfx
+    overflowPages = [ makeRawOverflowPageBS sfxpg
+                    | sfxpg <- takeWhile (not . BS.null)
+                                 [ BS.drop n sfx | n <- [0, 4096 .. ] ]
+                    ]
```
Comment (jorisdral) [inline test/Test/Util/RawPage.hs:36]: This looks like it doesn't satisfy the `makeRawOverflowPage` invariant that the length of `sfxpg` be 4096 or less
```suggestion
    overflowPages = [ makeRawOverflowPageBS sfxpg
                    | BS.take 4096 sfxpg <- takeWhile (not . BS.null)
                                 [ BS.drop n sfx | n <- [0, 4096 .. ] ]
                    ]
```
AFAICS it's not caught in a test -- maybe because the generated pages are too small?

File: test/Test/Util/RawPage.hs [hunk 9]
```diff
@@ -22,12 +24,26 @@ import           Test.Tasty.HUnit (Assertion, assertFailure)
...
+
+makeRawPageBS :: BS.ByteString -> RawPage
+makeRawPageBS bs =
+    case SBS.toShort bs of
+      SBS.SBS ba -> makeRawPage (ByteArray ba) 0
+
+makeRawOverflowPageBS :: BS.ByteString -> RawOverflowPage
+makeRawOverflowPageBS bs =
+    case SBS.toShort bs of
+      SBS.SBS ba -> makeRawOverflowPage (ByteArray ba) 0 (BS.length bs)
```
Comment (jorisdral) [inline test/Test/Util/RawPage.hs:46]: Or, alternative for my previous comment:
```suggestion
      SBS.SBS ba -> makeRawOverflowPage (ByteArray ba) 0 4096
```

**Review by jorisdral — COMMENTED** [review]
I haven't looked at all the tests, but here are some comments

File: test/Test/Database/LSMTree/Internal/RawPage.hs (hunk 6 above)
Comment (dcoutts) [reply test/Test/Database/LSMTree/Internal/RawPage.hs:268]: I've made this use clearer utils:
```
-- | Check that an 'Operation' and optional 'Blobref' (from the reference
-- implementation) is equivalent to (from the real implementation) the
-- combination of an 'Entry', containing the prefix of a value, plus a given
-- length suffix from the overflow pages.
equivalentOpEntry :: Operation -> Maybe BlobRef
                  -> Entry SerialisedValue BlobSpan
                  -> BS.ByteString
                  -> Int
                  -> Property

-- | Check that a value is equality to the combination of a 'SerialisedValue'
-- prefix, plus a given length of suffix from the overflow pages.
--
equivalentValue :: BS.ByteString -> Value -> SerialisedValue -> Int -> Property

equivalentBlobRefSpan :: BlobRef -> BlobSpan -> Property
```

File: src/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    | otherwise
+    = makeRawOverflowPageCopy ba off len
+
+makeRawOverflowPageCopy
+    :: ByteArray  -- ^ bytearray
+    -> Int        -- ^ offset in bytes into the bytearray
+    -> Int        -- ^ length in bytes
+    -> RawOverflowPage
+makeRawOverflowPageCopy ba off len =
+    assert (len >= 0 && len < 4096 && len <= sizeofByteArray ba - off) $
```
Comment (mheinzel) [inline src/Database/LSMTree/Internal/RawOverflowPage.hs:91]: Would this function not also be used on inputs where `len == 4096` (failing this invariant), but the byte array is not pinned?

```suggestion
    assert (len >= 0 && len <= 4096 && len <= sizeofByteArray ba - off) $
```

File: src/Database/LSMTree/Internal/RawOverflowPage.hs [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+    , isByteArrayPinned ba                -- bytearray must be pinned
+    ]
+
+instance NFData RawOverflowPage where
+  rnf (RawOverflowPage _ _) = ()
+
+-- | This instance assumes pages are 4096 bytes in size
+instance Eq RawOverflowPage where
+    RawOverflowPage off1 ba1 == RawOverflowPage off2 ba2 =
+      P.Vector off1 4096 ba1 == (P.Vector off2 4096 ba2 :: P.Vector Word8)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:62]: done.

File: src/Database/LSMTree/Internal/RawOverflowPage.hs (hunk 7 above)
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:155]: I've changed it to assert on the result rather than input. The real condition we want to check is that the conversion to overflow pages now works without copying. So we assert that.

File: src/Database/LSMTree/Internal/RawOverflowPage.hs (hunk 10 above)
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs:91]: Yes I agree. Done.

File: test/Test/Util/RawPage.hs (hunk 8 above)
Comment (dcoutts) [reply test/Test/Util/RawPage.hs:36]: I've tweaked the size distribution of the prototype's `LogicalPage` so that it generates bigger pages with the default QC size of 100. Previously it was generating bigger pages but only with size 200+.

And then yes it does detect the bug above, also fixed.

File: test/Test/Util/RawPage.hs (hunk 9 above)
Comment (dcoutts) [reply test/Test/Util/RawPage.hs:46]: took the other approach

File: src/Database/LSMTree/Internal/RawOverflowPage.hs (hunk 2 above)
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/RawOverflowPage.hs]: I *think* this is now resolved.

**Comment by dcoutts** [comment]
Ok, I think all review comments are now addressed.

**Comment by dcoutts** [comment]
I need to
1. fixup stylish-haskell complaints
2. add a `sameByteArray#` (in terms of `sameMutableByteArray#`) for ghc 9.2 and earlier, using CPP

**Review by mheinzel — APPROVED** [review]
LGTM, I'm fine with merging this once the shrinker and the issues you pointed out yourself are addressed.

File: src/utils/Database/LSMTree/Generators.hs [hunk 12]
```diff
@@ -574,6 +590,13 @@ instance Arbitrary SerialisedBlob where
   arbitrary = SerialisedBlob <$> genRawBytes
   shrink (SerialisedBlob rb) = SerialisedBlob <$> shrinkRawBytes rb
 
+newtype LargeRawBytes = LargeRawBytes RawBytes
+  deriving Show
+
+instance Arbitrary LargeRawBytes where
+  arbitrary = genRawBytesSized (4096*3) >>= fmap LargeRawBytes . genSlice
+  shrink (LargeRawBytes rb) = map LargeRawBytes (shrink rb)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/utils/Database/LSMTree/Generators.hs:598]: So `LargeRawBytes` is just a wrapper around `RawBytes`, so the right place to fix this would be in the `RawBytes` shrinker.

That one uses
```
  shrink rb = shrinkRawBytes rb ++ shrinkSlice rb
```
where `shrinkRawBytes` does the list thing and `shrinkSlice` uses a combo of take and drop.

So presumably we should just reverse the order of these two shrinkers and put the slice one first.

File: src/Database/LSMTree/Internal/BitMath.hs [hunk 13]
```diff
@@ -150,3 +152,12 @@ mul64 x = unsafeShiftL x 6
...
 {-# INLINE ceilDiv64 #-}
+
+-------------------------------------------------------------------------------
+-- 4096
+-------------------------------------------------------------------------------
+
+roundUpToPageSize :: Int -> Int
+roundUpToPageSize n =
+    -- assumes pageSize = 4096:
+    ((n + 0x0fff) .&. complement 0x0fff)
```
[the remark this answers is not in the document]
Comment (dcoutts) [reply src/Database/LSMTree/Internal/BitMath.hs]: :+1:

File: src/utils/Database/LSMTree/Generators.hs (hunk 12 above)
[the remark this answers is not in the document]
Comment (mheinzel) [reply src/utils/Database/LSMTree/Generators.hs:598]: The values generated by `Arbitrary RawBytes` should not be as large, but I agree, it's better to address the issue there.

Reversing the order is good, but doesn't entirely prevent the problem. If a property only fails for large (e.g. 4k+) `RawBytes`, once the shrinker reduced the length to that, it will iterate through all the list shrinks of a 4k+ elements list.

**Review by mheinzel — APPROVED** [review]
As noted, I think the shrinker can be improved further, but not a strict blocker from my side.
